### PR TITLE
support revealing defined opaque post borrowck

### PR DESCRIPTION
**Pull request by lcnr**
By adding a new `TypingMode::PostBorrowckAnalysis`. Currently only supported with the new solver and I didn't look into the way we replace `ReErased`. @compiler-errors mentioned that always using existentials may be unsound.

Checking that opaque types are well-formed has to be able to reveal other opaque types defined by the same scope. We previously did not properly support this and instead simply allowed `check_opaque_meets_bounds` to define these opaques again, and then checked that the resulting constraints match the hidden types of the defining function. This results in ambiguity errors when trying to normalize such an opaque without actually defining it. See https://github.com/rust-lang/rust/blob/master/tests/ui/traits/next-solver/opaques/no-define-in-wf-check.rs for examples. 

r? @compiler-errors

### CODE REVIEW

File: compiler/rustc_type_ir/src/interner.rs [hunk 1]
```diff
@@ -297,6 +297,12 @@ pub trait Interner:
         self,
         defining_anchor: Self::LocalDefId,
     ) -> Self::DefiningOpaqueTypes;
+
+    fn fold_regions<T: TypeFoldable<Self>>(
```
Comment (compiler-errors) [inline compiler/rustc_type_ir/src/interner.rs:301]: I feel like we should just uplift this folder. Should be possible to do.

File: compiler/rustc_trait_selection/src/traits/query/normalize.rs [hunk 2]
```diff
@@ -216,9 +216,9 @@ impl<'a, 'tcx> FallibleTypeFolder<TyCtxt<'tcx>> for QueryNormalizer<'a, 'tcx> {
             ty::Opaque => {
                 // Only normalize `impl Trait` outside of type inference, usually in codegen.
                 match self.infcx.typing_mode() {
-                    TypingMode::Coherence | TypingMode::Analysis { defining_opaque_types: _ } => {
-                        ty.try_super_fold_with(self)?
-                    }
+                    TypingMode::Coherence
+                    | TypingMode::Analysis { defining_opaque_types: _ }
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/query/normalize.rs:220]: nit: use `..` here?

File: compiler/rustc_trait_selection/src/traits/normalize.rs [hunk 3]
```diff
@@ -213,9 +214,10 @@ impl<'a, 'b, 'tcx> TypeFolder<TyCtxt<'tcx>> for AssocTypeNormalizer<'a, 'b, 'tcx
             ty::Opaque => {
                 // Only normalize `impl Trait` outside of type inference, usually in codegen.
                 match self.selcx.infcx.typing_mode() {
-                    TypingMode::Coherence | TypingMode::Analysis { defining_opaque_types: _ } => {
-                        ty.super_fold_with(self)
-                    }
+                    // FIXME(#132279): We likely want to reveal opaques during post borrowck analysis
+                    TypingMode::Coherence
+                    | TypingMode::Analysis { defining_opaque_types: _ }
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/normalize.rs:219]: here too, and elsewhere

**Review by compiler-errors — COMMENTED** [review]
Generally happy w/ the approach. I like how small the diff is given the typing mode refactor :D 

Could you try to uplift `fold_regions`? It feels bad to have that not be a folder defined in type_ir since it's really pretty simple otherwise.

File: compiler/rustc_trait_selection/src/traits/query/normalize.rs (hunk 2 above)
Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/query/normalize.rs:220]: Or change the `TypingMode::PostBorrowck` to list its fields exhaustively

**Comment by lcnr** [comment]
rdy

**Comment by lcnr** [comment]
actually rdy xd

**Review by compiler-errors — APPROVED** [review]

**Pull request merged**